### PR TITLE
Feature/husky setup tsc fix

### LINKED ISSUE #154: The CI should build + test the app to resolve build errors at the PR level - starting locally
I have had 3 consecutive build errors that were marked error-free by the current CI and approved for merging as a PR. The functionality of the CI should be to detect all build errors. I would once again recommend using Vercel in reference to #42 as it not only builds the app for every PR but serves it on a unique URL too. For the time being we should build the app on local before making PRs

### LINKED ISSUE #347: Make TS checking stricter and fix errors / fulfil type requirements
The whole purpose of integrating TS into the project was to block the commits/pushes of the app when the changes are not strongly typed. The current state of `tsconfig.json` doesn't show errors on the app but only displays the problems as `eslint` / `ts` warnings (an example is attached below).
There are two ways of doing this - 
1. Add TS compilation to the `npm run dev` script that starts the Vite server (This is the most used way)
2. Add a pre-commit hook that runs TS compilation (This is IMO better as it makes development smoother when commits have a small amount of code, but if the commits are big then you will have a ton of errors at the time of a commit and this might result in missed deadlines)

Please discuss below the preference for this and implement it ASAP. This was supposed to be done in #237.

Also attached below are the screencaps of various parts of the codebase showing that a lot of TS errors have been straight-up ignored. This is not good for the long run and we need to follow the convention and coding standards to make it easier for others to code and write efficient tests.

Screenshot (10)

Screenshot (11)

Screenshot (12)

Screenshot (13)

Screenshot (14)

Screenshot (15)

### PR DESCRIPTION
**Pull request by pulkitvyas08**
Resolves #347
Resolves #154

Adds husky pre-commit hook for linting the code on every commit so we don't waste time on spaces and semicolons during code reviews (please make sure that you have the prettier plugin installed on VS code though).

Adds a husky pre-push hook to check the validity of the code for typescript type usage and errors that the developer has to resolve before being able to push.

We haven't been doing `tsc` in the project till now which lets a lot of typescript errors go without catching them.

As part of this PR, I have tried to fix a lot of the typescript errors present in the codebase and suppressed the ones that were taking a lot of time to fix (as this isn't really in the scope of the PR)

#413 introduced changes that resulted in conflicts with how source maps generated by `tsc` work and how the build process of Vite works. So I had to play with some `tsconfig` options to arrive at the optimal case that doesn't give build errors when this script is executed: `tsc && vite build`. This took a lot of time, hence the late PR.

Hopefully, this makes devex much better. Mandated local testing can now be implemented quite easily just by extending the work of the pre-commit hook.

### CODE REVIEW

**Review by tijlleenders — CHANGES_REQUESTED** [review]
This is going to save time and frustration!
Haven't tested it but I do see this:
The /built folder is being included in the last commit. I think that's an accident?

**Comment by tijlleenders** [comment]
Does using husky.sh mean no Windows-based development?

File: cypress/integration/dashboard.spec.ts [hunk 1]
```diff
@@ -1,54 +1,66 @@
...
-    cy.get('.add-icon').should('have.length', 2);
+  it("Should have option to add Goals and Feelings", () => {
+    cy.get(".add-icon").should("have.length", 2);
     // cy.get('.add-icon').eq(0).click().location('pathname')
     // .should('include', 'MyGoals')
     // .go('back');
-    cy.get('.add-icon').eq(1).click().location('pathname')
-      .should('include', 'AddFeelings')
-      .go('back');
+    cy.get(".add-icon").eq(1).click().location("pathname").should("include", "AddFeelings").go("back");
```
Comment (Tushar-4781) [inline cypress/integration/dashboard.spec.ts:16]: Expected line break before .should. ( eslint )
Did husky skipped or missed it ?

File: cypress/integration/zinzen-page.spec.ts [hunk 2]
```diff
@@ -1,24 +1,28 @@
...
+        "https://www.gofundme.com/f/deliver-purpose-with-an-app-for-keeping-promises/donate"
+      );
   });
 
-  it('Should have Feedback btn that redirects to feedback page', () => {
-    cy.contains('Feedback').parent().click().location('pathname')
-      .should('include', 'Feedback')
-      .go('back');
+  it("Should have Feedback btn that redirects to feedback page", () => {
+    cy.contains("Feedback").parent().click().location("pathname").should("include", "Feedback").go("back");
```
Comment (Tushar-4781) [inline cypress/integration/zinzen-page.spec.ts:22]: here too Expected line break before .should. ( eslint )

File: i18next-parser.config.js [hunk 3]
```diff
@@ -1,27 +1,25 @@
...
-  }
+module.exports = {
+  createOldCatalogs: true, // save previous translation catalogs to the \_old folder
+
+  lexers: {
+    js: ["JsxLexer"], // we're writing jsx inside .js files
+    default: ["JavascriptLexer"],
+  },
+
+  locales: ["es", "es", "fr", "hi"],
```
Comment (Tushar-4781) [inline i18next-parser.config.js:9]: are these language options which we ask at the beginning ? if yes why there are only 3 languages? I think we have 7

**Review by Tushar-4781 — CHANGES_REQUESTED** [review]

**Comment by pulkitvyas08** [comment]
> This is going to save time and frustration! Haven't tested it but I do see this: The /built folder is being included in the last commit. I think that's an accident?

Yeah let me add that to ignore file

**Comment by pulkitvyas08** [comment]
> Does using husky.sh mean no Windows-based development?

No, I'm developing on windows, it works

File: cypress/integration/dashboard.spec.ts (hunk 1 above)
Comment (pulkitvyas08) [reply cypress/integration/dashboard.spec.ts:16]: Husky is now setup for prettier and not eslint anymore.

File: cypress/integration/zinzen-page.spec.ts (hunk 2 above)
Comment (pulkitvyas08) [reply cypress/integration/zinzen-page.spec.ts:22]: ESLint rules are too stringent and useless, ignore them or add the rules to ignore list

File: i18next-parser.config.js (hunk 3 above)
Comment (pulkitvyas08) [reply i18next-parser.config.js:9]: Not in the scope of this PR. Make a new issue to address this

**Review by Tushar-4781 — APPROVED** [review]

**Review by tijlleenders — APPROVED** [review]

**Pull request merged**